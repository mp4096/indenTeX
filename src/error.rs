use ignore;
use std::fmt;
use std::io;

#[derive(Debug)]
pub enum IndentexError {
    Io(io::Error),
    InvalidExtension,
    WalkError(ignore::Error),
    TranspileError,
}

impl From<ignore::Error> for IndentexError {
    fn from(e: ignore::Error) -> IndentexError {
        IndentexError::WalkError(e)
    }
}

impl From<io::Error> for IndentexError {
    fn from(e: io::Error) -> IndentexError {
        IndentexError::Io(e)
    }
}

impl fmt::Display for IndentexError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use std::error::Error;
        match *self {
            IndentexError::Io(ref e) => write!(f, "{}", e.description()),
            IndentexError::InvalidExtension => write!(f, "not a valid indentex file"),
            IndentexError::WalkError(ref e) => write!(f, "{}", e.description()),
            IndentexError::TranspileError => write!(f, "invalid indentex text"),
        }
    }
}
