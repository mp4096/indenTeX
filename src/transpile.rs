use std::path::Path;
use std::vec::Vec;
use error::IndentexError;


const LINESEP: &'static str = "\n";
const LATEX_TO_INDENTEX_FACTOR: f64 = 1.5;


// Indentation processing
#[inline]
fn count_left_indent<T: AsRef<str>>(line: T) -> Option<usize> {
    if line.as_ref().is_empty() {
        None
    } else {
        Some(line.as_ref().chars().count() - line.as_ref().trim_left().chars().count())
    }
}

fn scan_indents<T: AsRef<str>>(lines: &[T]) -> Vec<usize> {
    let raw_indents = lines.iter().map(count_left_indent).collect::<Vec<_>>();

    let mut adjusted_indents: Vec<usize> = Vec::with_capacity(raw_indents.len() + 1);
    let mut last_indent: usize = 0;

    for current_indent in raw_indents.iter().rev() {
        adjusted_indents.push(match *current_indent {
            None => last_indent,
            Some(ind) => {
                last_indent = ind;
                ind
            }
        });
    }

    adjusted_indents.reverse();
    adjusted_indents.push(0);

    adjusted_indents
}


// Transpilation
fn transpile<T: AsRef<str>>(lines: &[T]) -> String {
    use parsers::Environment;
    use parsers::Hashline::{PlainLine, OpenEnv};
    use parsers::process_line;

    // The number of environments is not known beforehand
    let mut env_stack: Vec<Environment> = Vec::new();

    // Input size is the sum of all line lengths plus the number of lines (for lineseps)
    let input_size = lines.iter().fold(0, |sum, l| sum + l.as_ref().len()) + lines.len();
    // We do not know how much larger the transpiled LaTeX file will be, but we can guess...
    let indentex_size = (LATEX_TO_INDENTEX_FACTOR * (input_size as f64)).round() as usize;
    let mut transpiled = String::with_capacity(indentex_size);

    let adjusted_indents = scan_indents(lines);

    for (line_num, line) in lines.iter().enumerate() {
        let list_like_active = match env_stack.last() {
            None => false, // No environment is active at all
            Some(d) => d.is_list_like(),
        };

        let tl = match process_line(line.as_ref(), list_like_active) {
            PlainLine(l) => l,
            OpenEnv(e) => {
                let tag_begin = e.latex_begin();
                env_stack.push(e);
                tag_begin
            }
        };
        transpiled.push_str(&tl);
        transpiled.push_str(LINESEP);

        // Check if we are in an environment and close as many as needed
        while match env_stack.last() {
            None => false,
            Some(d) => d.indent_depth() >= adjusted_indents[line_num + 1],
        } {
            // `unwrap()` is safe here since we have already checked if the stack is empty
            let tag_end = env_stack.pop().unwrap().latex_end();
            transpiled.push_str(&tag_end);
            transpiled.push_str(LINESEP);
        }
    }

    transpiled
}

pub fn transpile_file<T: AsRef<Path>>(path: T) -> Result<(), IndentexError> {
    use file_utils::{read_and_trim_lines, rename_indentex_file, write_to_file};

    let lines = read_and_trim_lines(path.as_ref())?;
    let transpiled_text = transpile(&lines);
    let path_out = rename_indentex_file(path)?;
    write_to_file(path_out, &transpiled_text)?;

    Ok(())
}


#[cfg(test)]
mod tests {
    #[test]
    fn count_left_indent() {
        use super::count_left_indent;

        assert_eq!(count_left_indent(""), None);
        assert_eq!(count_left_indent("foo"), Some(0));
        assert_eq!(count_left_indent("  bar"), Some(2));
    }
}
