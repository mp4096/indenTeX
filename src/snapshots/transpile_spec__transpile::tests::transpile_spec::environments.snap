---
created: "2019-06-16T17:28:50.217937894Z"
creator: insta@0.8.1
source: src/transpile.rs
expression: "transpile(input, &to)"
---
\begin{equation*}
  \label{eq:test}
  a + b
\end{equation*}

\begin{tikzpicture}[x = 2 cm]
  \draw (0, 0) -- (1, 1);
\end{tikzpicture}

Comments should be handled correctly:
\begin{equation} % test
  a + b
\end{equation}
\begin{remark}[test percent escaping \%] % baz
  foo bar
\end{remark}

Nested environment should be handled correctly
\begin{a}
  \begin{b}
    \begin{c}
      \begin{d}
        \begin{e}
          \begin{f}
            \begin{g}
              \begin{h}
                \begin{i}
                  foobar
                \end{i}
              \end{h}
            \end{g}
          \end{f}
        \end{e}
      \end{d}
    \end{c}
  \end{b}
\end{a}

